### ISSUE TITLE
scripts/basic/fixdep.c:410:1: fatal error: opening dependency file scripts/basic/.fixdep.d: Permission denied

### ISSUE THREAD

**Issue by afgcons** [AlexHowansky/lci_ldx9000_driver#2]
Asking for help, the SUBDIRS= in the Makefile is no longer supported some poster wrote who got the same errors compiling some video driver. My  uname -a
Linux 5.4.0-84-generic #94~18.04.1-Ubuntu SMP Thu Aug 26 23:17:46 UTC 2021 x86_64 GNU/Linux

Thanks for sharing and caring !

**Comment by AlexHowansky** [comment]
Welp, compiling on Fedora 34 doesn't work for me. That about exhausts the extent of my debugging ability with kernel modules. I'll poke around a bit, don't hold your breath. :)

**Comment by afgcons** [comment]
make -C /lib/modules/`uname -r`/build M=${PWD} modules

Found this about 15 min ago:
[https://stackoverflow.com/questions/35690076/simple-make-file-error](https://stackoverflow.com/questions/35690076/simple-make-file-error))

**Comment by afgcons** [comment]
###  cat Makefile
KERNELVERSION = `uname -r`
KERNELSRC := /lib/modules/$(KERNELVERSION)/build

obj-m += usblcpd.o

all: usblcpd.ko testlcpd

testlcpd: testlcpd.c
        gcc -o testlcpd testlcpd.c

usblcpd.ko: usblcpd.c
        $(MAKE) -I /usr/include -C $(KERNELSRC)  modules
        strip --strip-unneeded usblcpd.ko

clean:
        rm -f *.o *.ko *.mod.c .*.cmd Module.symvers modules.order testlcpd
        rm -rf .tmp_versions

install:
        mkdir -p /lib/modules/$(KERNELVERSION)/extra/drivers/usb/misc
        cp -v usblcpd.ko /lib/modules/$(KERNELVERSION)/extra/drivers/usb/misc
        depmod -a
        modprobe usblcpd


###  make -C /lib/modules/`uname -r`/build M=${PWD} modules
make: Entering directory '/usr/src/linux-headers-5.4.0-84-generic'
  CC [M]  /home/afg/Installed/lci_ldx9000_driver-master/usblcpd.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC [M]  /home/afg/Installed/lci_ldx9000_driver-master/usblcpd.mod.o
  LD [M]  /home/afg/Installed/lci_ldx9000_driver-master/usblcpd.ko
make: Leaving directory '/usr/src/linux-headers-5.4.0-84-generic'

-rw-rw-r--  1 afg afg 15122 2020-01-16 16:19 COPYING
-rw-rw-r--  1 afg afg   552 2021-10-05 15:46 Makefile
-rw-rw-r--  1 afg afg   568 2020-01-16 16:19 Makefile~
-rw-rw-r--  1 afg afg    57 2021-10-05 18:13 modules.order
-rw-rw-r--  1 afg afg     0 2021-10-05 18:13 Module.symvers
-rw-rw-r--  1 afg afg  1304 2020-01-16 16:19 README.md
-rw-rw-r--  1 afg afg  5083 2020-01-16 16:19 testlcpd.c
-rw-rw-r--  1 afg afg 10670 2020-01-16 16:19 usblcpd.c
-rw-rw-r--  1 afg afg 14088 2021-10-05 18:13 usblcpd.ko
-rw-rw-r--  1 afg afg   338 2021-10-05 18:13 .usblcpd.ko.cmd
-rw-rw-r--  1 afg afg    57 2021-10-05 18:13 usblcpd.mod
-rw-rw-r--  1 afg afg  1780 2021-10-05 18:13 usblcpd.mod.c
-rw-rw-r--  1 afg afg   199 2021-10-05 18:13 .usblcpd.mod.cmd
-rw-rw-r--  1 afg afg  4968 2021-10-05 18:13 usblcpd.mod.o
-rw-rw-r--  1 afg afg 31205 2021-10-05 18:13 .usblcpd.mod.o.cmd
-rw-rw-r--  1 afg afg 10024 2021-10-05 18:13 usblcpd.o
-rw-rw-r--  1 afg afg 41748 2021-10-05 18:13 .usblcpd.o.cmd

**Comment by afgcons** [comment]
ok, now it compiled but the install needs tweaking. the ..../extra/drivers/... doesn't exist. 
"/lib/modules/$(uname -r)/kernel/drivers/usb/misc/" looks like the best choice. 

Back to google I learn there is a [kbuild system](https://askubuntu.com/questions/1083091/how-to-compile-linux-kernel-module-printk-missing). I hope this helps!

**Comment by AlexHowansky** [comment]
Yep OK cool, I got it to the point where it compiles successfully, but missing the signing piece so it won't load. Still poking around.

**Comment by AlexHowansky** [comment]
OK pushed a new commit, please see if that works for you.

**Comment by afgcons** [comment]
The compile looks good but the install doesn't. That M="" doesn't look right. I will try to find an explanation. THANKS for helping!

### make
make -C /lib/modules/5.4.0-84-generic/build M="/home/afg/Installed/lci_ldx9000_driver"
make[1]: Entering directory '/usr/src/linux-headers-5.4.0-84-generic'
  AR      /home/afg/Installed/lci_ldx9000_driver/built-in.a
  CC [M]  /home/afg/Installed/lci_ldx9000_driver/usblcpd.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC [M]  /home/afg/Installed/lci_ldx9000_driver/usblcpd.mod.o
  LD [M]  /home/afg/Installed/lci_ldx9000_driver/usblcpd.ko
make[1]: Leaving directory '/usr/src/linux-headers-5.4.0-84-generic'
strip --strip-unneeded usblcpd.ko
gcc -o testlcpd testlcpd.c

### sudo make install
make -C /lib/modules/5.4.0-84-generic/build M="" modules_install
make[1]: Entering directory '/usr/src/linux-headers-5.4.0-84-generic'
sed: can't read modules.order: No such file or directory
Makefile:1379: recipe for target '_modinst_' failed
make[1]: *** [_modinst_] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.4.0-84-generic'
Makefile:16: recipe for target 'install' failed
make: *** [install] Error 2

**Comment by afgcons** [comment]
When I ran `sudo make install` it failed because M is a null string, yet the same `M="(PWD)"` expands as expected on the prior `make` step. Why I have no clue... If I put the sudo inside the Makefile and run `make install` or `make clean` (without sudo) the M variable contains $PWD. 

Then I got hit by kernel signing: "SSL error:02001002:system library:fopen:No such file or directory: ../crypto/bio/bss_file.c:72". [Here](https://github.com/Xilinx/dma_ip_drivers/issues/39) _Thesane_ offered a simple solution:
```
cd /usr/src/linux-headers-$(uname -r)/certs
sudo openssl req -new -x509 -newkey rsa:2048 -keyout signing_key.pem -outform DER -out signing_key.x509 -nodes -subj "/CN=Owner/"
```
Below is your patched up Makefile, containig bits I found [here](https://stackoverflow.com/questions/57521081/warning-modules-install-missing-system-map-file-skipping-depmod?newreg=1d93e8b4e35549299be4aa684f5b8d24). It works - in a non-standard way...
```
obj-m += usblcpd.o

# Default to running kernel's build directory if KDIR not set externally
KDIR ?= "/lib/modules/$(shell uname -r)/build"

all: usblcpd.ko testlcpd

testlcpd: testlcpd.c
	gcc -o testlcpd testlcpd.c

usblcpd.ko: usblcpd.c
	$(MAKE) -C "$(KDIR)" M="$(PWD)"
	strip --strip-unneeded usblcpd.ko

install: usblcpd.ko
	sudo $(MAKE) -C "$(KDIR)" M="$(PWD)" modules_install
	sudo depmod -A

clean:
	rm -f *.o *.ko *.mod *.mod.c .*.cmd Module.symvers modules.order built-in.* testlcpd
	sudo $(MAKE) -C "$(KDIR)" M="$(PWD)" clean
```
[Makefile.txt](https://github.com/AlexHowansky/lci_ldx9000_driver/files/7307022/Makefile.txt)

**Comment by AlexHowansky** [comment]
Ah PWD is empty because sudo wipes the environment of variables it doesn't want to pass to the child process. You should be able to replace `M="$(PWD)"` with `M=$(shell pwd)` to correct this. Not sure about the cert signing, it shouldn't be required unless you have explicitly enabled it. Does this process give you a .ko that you can successfully insmod? I can compile and sign but get `Invalid module format` when doing insmod with it.
